src/sampling_utils.py

```diff
@@ -0,0 +1,487 @@
+
+                # get list of day-ids for all county-days
+                dayids = np.arange(len(data.index))
+                day_of_cntyday = np.tile(dayids, len(data.columns))
+
+                # get list of day-ids for all samples
+                day_of_smpl = np.array([ day_of_cntyday[i] for (i,smpls) in enumerate(smpls_per_cntyday) for x in range(smpls) ]) # TODO: np.empty
+
+                # get available times for each sample
+                time_of_days = data.index.tolist() # cannot be a np.array as it needs to stay a pandas.timeformat
+                av_times_per_day = [len(times_by_day[d]) for d in time_of_days] #add '_np' to times_by_day
+                av_times_per_smpl = [ av_times_per_day[day_of_cntyday[i]] for (i,smpls) in enumerate(smpls_per_cntyday) for x in range(smpls) ]
+
+                ######## x_all ########
+
+                # get list of county-ids for all county-days
+                cntyids = np.arange(len(data.columns))
+                cnty_of_cntyday = np.repeat(cntyids, len(data.index))
+
+                # get list of county-ids for all samples
+                cnty_of_smpl = np.array([ cnty_of_cntyday[i] for (i,smpl) in enumerate(smpls_per_cntyday) for x in range(smpl) ])
+
+                # get available locations for each sample
+                label_of_cntys = data.columns # list of countys labels
+                av_locs_per_cnty = [len(locations_by_county[c]) for c in label_of_cntys] #add '_np' to locations_by_county
+                av_locs_per_smpl = [ av_locs_per_cnty[cnty_of_cntyday[i]] for (i,smpls) in enumerate(smpls_per_cntyday) for x in range(smpls) ]
+
+                return (n_total, dayoffset,
+                day_of_smpl, av_times_per_smpl,
+                cnty_of_smpl, av_locs_per_smpl)
+
+
+def sample_time_and_space__pred(n_days, n_counties, times_by_day_np, locations_by_county_np, d_offs, c_offs, num_tps, av_times_per_smpl, av_locs_per_smpl, rnd_time, rnd_loc):
+
+                    ######## t_all ########
+                    n_total = n_days * n_counties * num_tps
+
+                    rnd_timeid_per_smpl = np.floor( av_times_per_smpl * rnd_time.random( n_total ) ).astype("int32")
+
+                    # collect times for each sample with its random time-id
+                    t_all = [ times_by_day_np[d_offs+i][rnd_timeid_per_smpl[(i*n_counties+j)*num_tps+x]] for i in range(n_days) for j in range(n_counties) for x in range(num_tps) ]
+
+                    ######## x_all ########
+
+                    # calc random location-id for each sample
+                    rnd_locid_per_smpl = np.floor( av_locs_per_smpl * rnd_loc.random((n_total,)) ).astype("int32")
+
+                    # collect locations for each sample with its random location-id
+                    x_all = [ locations_by_county_np[c_offs+j][rnd_locid_per_smpl[(i*n_counties+j)*num_tps+x]] for i in range(n_days) for j in range(n_counties) for x in range(num_tps) ]
+
+                    return t_all, x_all

```

Comment:
too many tabs! please try to keep to a 90 characters per line limit where possible!

Reply:
Fixed it. Lines now should not exceed 60 characters